src/System.Linq.Expressions/src/System/Runtime/CompilerServices/ReadOnlyCollectionBuilder.cs

```diff
@@ -174,15 +199,21 @@ public void RemoveAt(int index)
         {
             get
             {
-                if (index >= _size)
+                // Following trick can reduce the range check by one comparison
+                if ((uint)index >= (uint)Count)
```

Comment:
> make such changes to produce correct Exceptions

The only real value in throwing better exceptions in cases like this is to help with debugging, and there's almost no difference between the diagnostic power of ArgumentOutOfRangeException and IndexOutOfRangeException.  So from my perspective, we're adding close to zero value while breaking consistency with desktop and potentially breaking (arguably bad) code that's catching and handling the IndexOutOfRangeException but not the ArgumentOutOfRangeException.